nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java

```diff
@@ -0,0 +1,399 @@
+     * that is smaller than the WriteBatcher's batch size will be flushed immediately and not have to wait for more
+     * FlowFiles to arrive to fill out the batch.
+     *
+     */
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            context.yield();
+            if (shouldFlushIfEmpty) {
+                flushWriteBatcherAsync(this.writeBatcher);
+            }
+            shouldFlushIfEmpty = false;
+        } else {
+            shouldFlushIfEmpty = true;
+
+            WriteEvent writeEvent = buildWriteEvent(context, session, flowFile);
+            if (getLogger().isDebugEnabled()) {
+                getLogger().debug("Writing URI: " + writeEvent.getTargetUri());
+            }
+            addWriteEvent(this.writeBatcher, writeEvent);
+        }
+    }
+
+    /*
+     * Protected so that it can be overridden for unit testing purposes.
+     */
+    protected void flushWriteBatcherAsync(WriteBatcher writeBatcher) {
+        writeBatcher.flushAsync();
+    }
+
+    /*
+     * Protected so that it can be overridden for unit testing purposes.
+     */
+    protected void addWriteEvent(WriteBatcher writeBatcher, WriteEvent writeEvent) {
+        writeBatcher.add(writeEvent);
+    }
+
+    protected WriteEvent buildWriteEvent(ProcessContext context, ProcessSession session, FlowFile flowFile) {
+        String uri = flowFile.getAttribute(context.getProperty(URI_ATTRIBUTE_NAME).getValue());
+        final String prefix = context.getProperty(URI_PREFIX).getValue();
+        if (prefix != null) {
+            uri = prefix + uri;
+        }
+        final String suffix = context.getProperty(URI_SUFFIX).getValue();
+        if (suffix != null) {
+            uri += suffix;
+        }
+
+        DocumentMetadataHandle metadata = new DocumentMetadataHandle();
+        final String collections = context.getProperty(COLLECTIONS).getValue();
+        if (collections != null) {
+            metadata.withCollections(collections.split(","));
+        }
+        final String permissions = context.getProperty(PERMISSIONS).getValue();
+        if (permissions != null) {
+            String[] tokens = permissions.split(",");
+            for (int i = 0; i < tokens.length; i += 2) {
+                String role = tokens[i];
+                String capability = tokens[i + 1];
+                metadata.withPermission(role, DocumentMetadataHandle.Capability.getValueOf(capability));
+            }
+        }
+        // Add the flow file UUID for Provenance purposes and for sending them
+        // to the appropriate relationship
+        String flowFileUUID = flowFile.getAttribute(CoreAttributes.UUID.key());
+        metadata.withMetadataValue("flowFileUUID", flowFileUUID);
+        // TODO Haven't had luck with wrapping the FlowFile's inputStream with an ML InputStreamHandle, so copying everything into a byte array
```

Comment:
recommend removal of all commented code/sections.  Could file a follow on JIRA for such actions if desired